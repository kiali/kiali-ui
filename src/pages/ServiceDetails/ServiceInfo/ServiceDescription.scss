.small-text-donut .donut-title-big-pf {
  font-size: 18px;
}

.small-text-donut .c3-chart-arc text {
  display: none;
}
