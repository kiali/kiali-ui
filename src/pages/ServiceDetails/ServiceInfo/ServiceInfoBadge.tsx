import * as React from 'react';

interface BadgeProps {
  scale: number;
  style: string;
  color: string;
  leftText: string;
  rightText: string;
  key?: string;
}

interface ServiceBadgeState {
  leftWidth: number;
  rightWidth: number;
}
class ServiceInfoBadge extends React.Component<BadgeProps, ServiceBadgeState> {
  private textLeftText: SVGTextContentElement | null;
  private textRightText: SVGTextContentElement | null;

  constructor(props: BadgeProps) {
    super(props);
    this.state = {
      leftWidth: 0,
      rightWidth: 0
    };
    this.updateTextLength = this.updateTextLength.bind(this);
  }
  componentDidMount() {
    this.updateTextLength();
  }
  componentDidUpdate() {
    this.updateTextLength();
  }
  updateTextLength() {
    let leftTextLength = this.TextLength(this.textLeftText);
    let rightTextLength = this.TextLength(this.textRightText);
    if (leftTextLength !== this.state.leftWidth || rightTextLength !== this.state.rightWidth) {
      this.setState({
        leftWidth: leftTextLength,
        rightWidth: rightTextLength
      });
    }
  }
  TextLength(node: SVGTextContentElement | null) {
    return Math.round(node ? node.getComputedTextLength() : 0);
  }
  render() {
    let minorMargin = this.props.scale;
    let majorMargin = 5 * this.props.scale;
    let textMargin = minorMargin + majorMargin;
    let fullWidth = textMargin * 4 + (this.state.leftWidth + this.state.rightWidth);
    let rightOffset = textMargin * 2 + this.state.leftWidth;
    let rightWidth = textMargin * 2 + this.state.rightWidth;
    let height = 20 * this.props.scale;
    let textSize = 11 * this.props.scale;
    let borderRadius = 3 * this.props.scale * (this.props.style === 'square' ? 0 : 1);
    let key = this.props.key || this.props.leftText + '_' + this.props.rightText;
    return (
      <svg
        height={height}
        width={fullWidth}
        zoomAndPan="magnify"
        preserveAspectRatio="xMidYMid meet"
        xmlns="http://www.w3.org/2000/svg"
        version="1.0"
        style={{ marginLeft: '5px' }}
        key={key}
      >
        <linearGradient id="a" x2="0" y2={this.props.style === 'plastic' ? '100%' : '0'}>
          <stop stopOpacity=".1" stopColor="#bbb" offset="0" />
          <stop stopOpacity=".1" offset="1" />
        </linearGradient>
        <rect rx={borderRadius} fill="#555" width={fullWidth} height={height} />
        <rect rx={borderRadius} fill={this.props.color} width={rightWidth} x={rightOffset} height={height} />
        <rect fill={this.props.color} x={rightOffset} width="13" height={height} />
        <rect rx={borderRadius} fill="url(#a)" width={fullWidth} height={height} />
        <g fontSize={textSize} fontFamily="DejaVu Sans,Verdana,Geneva,sans-serif" fill="#fff">
          <text
            ref={ref => (this.textLeftText = ref)}
            x={textMargin}
            fill="#010101"
            fillOpacity=".3"
            y={height - majorMargin}
          >
            {this.props.leftText}
          </text>
          <text x={textMargin} y={height - majorMargin - minorMargin}>
            {this.props.leftText}
          </text>
          <text
            ref={ref => (this.textRightText = ref)}
            fill="#010101"
            x={rightOffset + textMargin}
            fillOpacity=".3"
            y={height - majorMargin}
          >
            {this.props.rightText}
          </text>
          <text x={rightOffset + textMargin} y={height - majorMargin - minorMargin}>
            {this.props.rightText}
          </text>
        </g>
      </svg>
    );
  }
}

export default ServiceInfoBadge;
