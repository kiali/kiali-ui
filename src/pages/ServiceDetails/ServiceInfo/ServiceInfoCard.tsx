import * as React from 'react';
import { Icon } from 'patternfly-react';

interface CardProps {
  iconType?: string;
  iconName?: string;
  title?: string;
  items?: any;
}

class ServiceInfoCard extends React.Component<CardProps> {
  constructor(props: CardProps) {
    super(props);
  }

  render() {
    return (
      <div className="card-pf">
        <div className="card-pf-heading">
          <h2 className="card-pf-title">
            <Icon type={this.props.iconType} name={this.props.iconName} style={{ marginRight: '10px' }} />
            {this.props.title}
          </h2>
        </div>
        <div className="card-pf-body">{this.props.items}</div>
      </div>
    );
  }
}

export default ServiceInfoCard;
