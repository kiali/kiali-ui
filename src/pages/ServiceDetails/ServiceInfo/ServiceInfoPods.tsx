import * as React from 'react';
import { Col, Row, OverlayTrigger, Tooltip } from 'patternfly-react';
import { Pod } from '../../../types/IstioObjects';
import { PodsGroup, groupPods } from './ServiceInfoPodsGrouping';
import Label from '../../../components/Label/Label';

interface Props {
  pods?: Pod[];
}

interface ServiceInfoPodsState {
  groups: PodsGroup[];
}
class ServiceInfoPods extends React.Component<Props, ServiceInfoPodsState> {
  static getDerivedStateFromProps(props: Props, _currentState: ServiceInfoPodsState) {
    return { groups: ServiceInfoPods.updateGroups(props) };
  }

  static updateGroups(props: Props) {
    if (props.pods) {
      return groupPods(props.pods);
    } else {
      return [];
    }
  }

  constructor(props: Props) {
    super(props);
    this.state = { groups: ServiceInfoPods.updateGroups(props) };
  }

  render() {
    return (
      <div className="card-pf">
        <Row className="row-cards-pf">
          <Col xs={12} sm={12} md={12} lg={12}>
            {(this.state.groups || []).map((group, u) => (
              <div className="card-pf-body" key={'pods_' + u}>
                <h3>
                  {group.numberOfPods > 1 ? (
                    <OverlayTrigger
                      // Prettier makes irrelevant line-breaking clashing with tslint
                      // prettier-ignore
                      overlay={<Tooltip id={'pod_names_' + u} title="Pod Names">{group.names.join(', ')}</Tooltip>}
                      placement="top"
                      trigger={['hover', 'focus']}
                    >
                      <span>{group.commonPrefix + '... (' + group.numberOfPods + ' replicas)'}</span>
                    </OverlayTrigger>
                  ) : (
                    group.commonPrefix + ' (1 replica)'
                  )}
                </h3>
                <div key="labels" className="label-collection">
                  {Object.keys(group.commonLabels).map((key, i) => (
                    <Label key={'pod_' + u + '_' + i} name={key} value={group.commonLabels[key]} />
                  ))}
                </div>
                <div>
                  <span>
                    {group.createdAtStart === group.createdAtEnd ? (
                      <>
                        <strong>Created at: </strong>
                        {new Date(group.createdAtStart).toLocaleString()}
                      </>
                    ) : (
                      <>
                        <strong>Created between: </strong>
                        {new Date(group.createdAtStart).toLocaleString() +
                          ' and ' +
                          new Date(group.createdAtEnd).toLocaleString()}
                      </>
                    )}
                  </span>
                </div>
                {group.createdBy.length > 0 && (
                  <div>
                    <span>
                      <strong>Created by: </strong>
                      {group.createdBy.map(ref => ref.name + ' (' + ref.kind + ')').join(', ')}
                    </span>
                  </div>
                )}
                {group.istioInitContainers && (
                  <div>
                    <span>
                      <strong>Istio init containers: </strong>
                      {group.istioInitContainers.map(c => `${c.name} [${c.image}]`).join(', ')}
                    </span>
                  </div>
                )}
                {group.istioContainers && (
                  <div>
                    <span>
                      <strong>Istio containers: </strong>
                      {group.istioContainers.map(c => `${c.name} [${c.image}]`).join(', ')}
                    </span>
                  </div>
                )}
                <hr />
              </div>
            ))}
          </Col>
        </Row>
      </div>
    );
  }
}

export default ServiceInfoPods;
