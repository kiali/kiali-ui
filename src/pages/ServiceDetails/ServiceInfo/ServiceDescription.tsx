import * as React from 'react';
import { Card, CardBody, Stack, StackItem, Tab, Tabs, Title, Tooltip } from '@patternfly/react-core';
import { EyeIcon } from '@patternfly/react-icons';
import { style } from 'typestyle';
import LocalTime from '../../../components/Time/LocalTime';
import { Endpoints, ServicePort } from '../../../types/ServiceInfo';
import { ObjectCheck, ObjectValidation } from '../../../types/IstioObjects';
import { ValidationObjectSummary } from '../../../components/Validations/ValidationObjectSummary';
import ValidationList from '../../../components/Validations/ValidationList';
import Labels from '../../../components/Label/Labels';
import { AdditionalItem } from 'types/Workload';
import { TextOrLink } from 'components/TextOrLink';
import { renderAPILogo } from 'components/Logo/Logos';
import './ServiceDescription.css';
import MissingSidecar from '../../../components/MissingSidecar/MissingSidecar';

interface ServiceInfoDescriptionProps {
  name: string;
  namespace: string;
  createdAt: string;
  resourceVersion: string;
  additionalDetails: AdditionalItem[];
  istioEnabled?: boolean;
  labels?: { [key: string]: string };
  selectors?: { [key: string]: string };
  type?: string;
  ip?: any;
  externalName?: string;
  ports?: ServicePort[];
  endpoints?: Endpoints[];
  validations?: ObjectValidation;
}

type State = {
  serviceInfoTabKey: number;
};

const listStyle = style({
  listStyleType: 'none',
  padding: 0
});

const ExternalNameType = 'ExternalName';

class ServiceDescription extends React.Component<ServiceInfoDescriptionProps, State> {
  constructor(props: ServiceInfoDescriptionProps) {
    super(props);
    this.state = {
      serviceInfoTabKey: 0
    };
  }

  serviceInfoHandleTabClick = (_event, tabIndex) => {
    this.setState({
      serviceInfoTabKey: tabIndex
    });
  };

  getPortOver(portId: number) {
    return <ValidationList checks={this.getPortChecks(portId)} />;
  }

  getPortChecks(portId: number): ObjectCheck[] {
    return this.props.validations
      ? this.props.validations.checks.filter(c => c.path === 'spec/ports[' + portId + ']')
      : [];
  }

  hasIssue(portId: number): boolean {
    return this.getPortChecks(portId).length > 0;
  }

  render() {
    return (
      <Card>
        <CardBody>
          <Title headingLevel="h3" size="2xl">
            {' '}
            Service{' '}
          </Title>
          <Tabs
            isFilled={true}
            activeKey={this.state.serviceInfoTabKey}
            onSelect={this.serviceInfoHandleTabClick}
            style={{ marginTop: '20px' }}
          >
            <Tab eventKey={0} title={'Properties'}>
              <Stack gutter={'md'} style={{ marginTop: '20px' }}>
                <StackItem id="name">
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Name{' '}
                  </Title>
                  {this.props.name}
                  {!this.props.istioEnabled && (
                    <span style={{ marginLeft: '10px' }}>
                      <MissingSidecar namespace={this.props.namespace} />
                    </span>
                  )}
                </StackItem>
                <StackItem id={'labels'}>
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Labels{' '}
                  </Title>
                  <Labels labels={this.props.labels || {}} />
                </StackItem>
                <StackItem id={'resource_version'}>
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Resource Version{' '}
                  </Title>
                  {this.props.resourceVersion}
                </StackItem>
                <StackItem id={'selectors'}>
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Selectors{' '}
                  </Title>
                  <Labels labels={this.props.selectors || {}} />
                </StackItem>
                <StackItem id={'created_at'}>
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Created at{' '}
                  </Title>
                  <LocalTime time={this.props.createdAt} />
                </StackItem>
                {this.props.additionalDetails.map((additionalItem, idx) => {
                  return (
                    <StackItem key={'additional-details-' + idx} id={'additional-details-' + idx}>
                      <Title headingLevel="h6" size="md">
                        {' '}
                        {additionalItem.title}{' '}
                      </Title>
                      {additionalItem.icon && renderAPILogo(additionalItem.icon, undefined, idx)}
                      <TextOrLink text={additionalItem.value} urlTruncate={64} />
                    </StackItem>
                  );
                })}
              </Stack>
            </Tab>
            <Tab eventKey={1} title={'Network'}>
              <Stack gutter={'md'} style={{ marginTop: '20px' }}>
                <StackItem id={'ip'}>
                  <Title headingLevel="h6" size="md">
                    {' '}
                    {this.props.type !== ExternalNameType ? 'Service IP' : 'ExternalName'}{' '}
                  </Title>
                  {this.props.type !== ExternalNameType
                    ? this.props.ip
                      ? this.props.ip
                      : ''
                    : this.props.externalName
                    ? this.props.externalName
                    : ''}
                </StackItem>
                <StackItem id={'endpoints'}>
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Endpoints{' '}
                  </Title>
                  <Stack gutter={'md'}>
                    {(this.props.endpoints || []).map((endpoint, i) =>
                      (endpoint.addresses || []).map((address, u) => (
                        <StackItem key={'endpoint_' + i + '_address_' + u}>
                          {address.name !== '' ? (
                            <Tooltip content={<>{address.name}</>}>
                              <span>
                                <EyeIcon /> {address.ip}
                              </span>
                            </Tooltip>
                          ) : (
                            <>{address.name}</>
                          )}
                        </StackItem>
                      ))
                    )}
                  </Stack>
                </StackItem>
                <StackItem id={'ports'}>
                  <Title headingLevel="h6" size="md">
                    <ValidationObjectSummary
                      id={this.props.name + '-config-validation'}
                      validations={this.props.validations ? [this.props.validations] : []}
                    />
                    <span style={{ marginLeft: '10px' }}>Ports</span>
                  </Title>
                  <ul className={listStyle}>
                    {(this.props.ports || []).map((port, i) => (
                      <li key={'port_' + i}>
                        {this.hasIssue(i) ? this.getPortOver(i) : undefined} {port.protocol} {port.name} ({port.port})
                      </li>
                    ))}
                  </ul>
                </StackItem>
                <StackItem id={'type'}>
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Type{' '}
                  </Title>
                  {this.props.type ? this.props.type : ''}
                </StackItem>
              </Stack>
            </Tab>
          </Tabs>
        </CardBody>
      </Card>
    );
  }
}

export default ServiceDescription;
