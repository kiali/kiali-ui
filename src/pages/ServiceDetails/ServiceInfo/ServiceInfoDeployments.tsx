import * as React from 'react';
import { Col, Row, Icon } from 'patternfly-react';
import LocalTime from '../../../components/Time/LocalTime';
import Badge from '../../../components/Badge/Badge';
import { Deployment } from '../../../types/ServiceInfo';
import { ratioCheck, Status } from '../../../components/ServiceHealth/HealthHelper';
import { PfColors } from '../../../components/Pf/PfColors';

interface ServiceInfoDeploymentsProps {
  deployments?: Deployment[];
}

class ServiceInfoDeployments extends React.Component<ServiceInfoDeploymentsProps> {
  constructor(props: ServiceInfoDeploymentsProps) {
    super(props);
  }

  render() {
    return (
      <div className="card-pf">
        <Row className="row-cards-pf">
          <Col xs={12} sm={12} md={12} lg={12}>
            {(this.props.deployments || []).map((deployment, u) => (
              <div className="card-pf-body" key={'deployments_' + u}>
                <h3>{deployment.name}</h3>
                <div key="labels">
                  {Object.keys(deployment.labels || new Map()).map((key, i) => (
                    <Badge
                      key={'deployment_' + i}
                      scale={0.8}
                      style="plastic"
                      color={PfColors.Green400}
                      leftText={key}
                      rightText={deployment.labels ? deployment.labels[key] : ''}
                    />
                  ))}
                </div>
                <div>
                  <strong>Pod status: </strong> {deployment.available_replicas} / {deployment.replicas}{' '}
                  {this.renderStatus(ratioCheck(deployment.available_replicas, deployment.replicas))}
                </div>
                {deployment.autoscaler.name !== '' && (
                  <div>
                    <strong>Autoscaler: </strong>
                    from {deployment.autoscaler.min_replicas} to {deployment.autoscaler.max_replicas} pods ({
                      deployment.autoscaler.target_cpu_utilization_percentage
                    }% CPU)
                  </div>
                )}
                <div>
                  <span>
                    <strong>Created at: </strong>
                    <LocalTime time={deployment.created_at} />
                  </span>
                </div>
                <div>
                  <span>
                    <strong>Resource Version: </strong>
                    {deployment.resource_version}
                  </span>
                </div>
                <hr />
              </div>
            ))}
          </Col>
        </Row>
      </div>
    );
  }

  renderStatus(status: Status) {
    if (status.icon) {
      return <Icon type="pf" name={status.icon} />;
    } else {
      return <span style={{ color: status.color }}>{status.text}</span>;
    }
  }
}

export default ServiceInfoDeployments;
