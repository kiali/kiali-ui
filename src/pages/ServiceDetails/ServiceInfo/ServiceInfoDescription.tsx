import * as React from 'react';
import { Col, Row } from 'patternfly-react';
import LocalTime from '../../../components/Time/LocalTime';
import { DisplayMode, HealthIndicator } from '../../../components/Health/HealthIndicator';
import { ServiceHealth } from '../../../types/Health';
import { Endpoints } from '../../../types/ServiceInfo';
import { Port, ObjectValidation } from '../../../types/IstioObjects';
import { style } from 'typestyle';
import {
  ConfigIndicator,
  NOT_VALID,
  SMALL_SIZE,
  MEDIUM_SIZE
} from '../../../components/ConfigValidation/ConfigIndicator';
import { Popover, OverlayTrigger, Icon, Tooltip } from 'patternfly-react';
import './ServiceInfoDescription.css';
import Labels from '../../../components/Label/Labels';
import { ThreeScaleServiceRule } from '../../../types/ThreeScale';

interface ServiceInfoDescriptionProps {
  name: string;
  namespace: string;
  createdAt: string;
  resourceVersion: string;
  istioEnabled?: boolean;
  labels?: { [key: string]: string };
  selectors?: { [key: string]: string };
  type?: string;
  ip?: any;
  externalName?: string;
  ports?: Port[];
  endpoints?: Endpoints[];
  health?: ServiceHealth;
  threeScaleServiceRule?: ThreeScaleServiceRule;
  validations?: ObjectValidation;
}

const listStyle = style({
  listStyleType: 'none',
  padding: 0
});

const labelTitleStyle = style({
  marginBottom: '2px'
});

const labelListStyle = style({
  marginBottom: '4px'
});

const ExternalNameType = 'ExternalName';

class ServiceInfoDescription extends React.Component<ServiceInfoDescriptionProps> {
  getValidations(): ObjectValidation {
    return this.props.validations ? this.props.validations : ({} as ObjectValidation);
  }

  getPortOver(portId: number): Popover {
    return (
      <Popover id={portId + '-config-validation'} title={NOT_VALID.name} style={{ maxWidth: '80%', minWidth: '200px' }}>
        <div>{this.getPortIssue(portId)}</div>
      </Popover>
    );
  }

  getPortIssue(portId: number): string {
    let message = '';
    if (this.props.validations && this.props.validations.checks) {
      message = this.props.validations.checks
        .filter(c => c.path === 'spec/ports[' + portId + ']')
        .map(c => c.message)
        .join(',');
    }
    return message;
  }

  hasIssue(portId: number): boolean {
    return this.getPortIssue(portId) !== '';
  }

  render() {
    return (
      <div className="card-pf">
        <div className="card-pf-body">
          <Row>
            <Col xs={12} sm={6} md={5} lg={5}>
              <div id="labels">
                <div className={'progress-description ' + labelTitleStyle}>
                  <strong>Labels</strong>
                </div>
                <div className={'label-collection ' + labelListStyle}>
                  <Labels labels={this.props.labels || {}} />
                </div>
              </div>
              <div id="selectors">
                <div className={'progress-description ' + labelTitleStyle}>
                  <strong>Selectors</strong>
                </div>
                <div className={'label-collection ' + labelListStyle}>
                  <Labels labels={this.props.selectors || {}} />
                </div>
              </div>
              <div>
                <strong>Type</strong> {this.props.type ? this.props.type : ''}
              </div>
              {this.props.type !== ExternalNameType ? (
                <div>
                  <strong>IP</strong> {this.props.ip ? this.props.ip : ''}
                </div>
              ) : (
                <div>
                  <strong>ExternalName</strong> {this.props.externalName ? this.props.externalName : ''}
                </div>
              )}
              <div>
                <strong>Created at</strong> <LocalTime time={this.props.createdAt} />
              </div>
              <div>
                <strong>Resource Version</strong> {this.props.resourceVersion}
              </div>
              {this.props.threeScaleServiceRule && this.props.threeScaleServiceRule.threeScaleHandlerName !== '' && (
                <span>
                  Service linked with 3scale API Handler <i>{this.props.threeScaleServiceRule.threeScaleHandlerName}</i>
                </span>
              )}
            </Col>
            <Col xs={12} sm={4} md={2} lg={2}>
              <div className="progress-description">
                <ConfigIndicator
                  id={this.props.name + '-config-validation'}
                  validations={[this.getValidations()]}
                  size={MEDIUM_SIZE}
                />
                <strong>Ports</strong>
              </div>
              <ul className={listStyle}>
                {(this.props.ports || []).map((port, i) => (
                  <li key={'port_' + i}>
                    {this.hasIssue(i) ? (
                      <OverlayTrigger
                        placement={'right'}
                        overlay={this.getPortOver(i)}
                        trigger={['hover', 'focus']}
                        rootClose={false}
                      >
                        <span style={{ color: NOT_VALID.color }}>
                          <Icon
                            type="pf"
                            name="error-circle-o"
                            style={{ fontSize: SMALL_SIZE }}
                            className="health-icon"
                            tabIndex="0"
                          />
                        </span>
                      </OverlayTrigger>
                    ) : (
                      undefined
                    )}
                    {port.protocol} {port.name} ({port.port})
                  </li>
                ))}
              </ul>
            </Col>
            <Col xs={12} sm={6} md={2} lg={2}>
              <div className="progress-description">
                <strong>Endpoints</strong>
              </div>
              {(this.props.endpoints || []).map((endpoint, i) => (
                <Row key={'endpoint_' + i}>
                  <Col xs={12} sm={12} md={12} lg={12}>
                    <ul className={listStyle}>
                      {(endpoint.addresses || []).map((address, u) => {
                        const id = 'endpoint_' + i + '_address_' + u;
                        return (
                          <li key={id}>
                            <OverlayTrigger
                              overlay={<Tooltip id={id + '_tooltip'}>{address.name}</Tooltip>}
                              trigger={['hover', 'focus']}
                              rootClose={false}
                            >
                              <strong>{address.ip} </strong>
                            </OverlayTrigger>
                          </li>
                        );
                      })}
                    </ul>
                  </Col>
                </Row>
              ))}
            </Col>
            <Col xs={12} sm={6} md={3} lg={3}>
              <div className="progress-description">
                <strong>Health</strong>
              </div>
              <HealthIndicator
                id={this.props.name}
                health={this.props.health}
                mode={DisplayMode.LARGE}
                tooltipPlacement="left"
              />
            </Col>
          </Row>
        </div>
      </div>
    );
  }
}

export default ServiceInfoDescription;
