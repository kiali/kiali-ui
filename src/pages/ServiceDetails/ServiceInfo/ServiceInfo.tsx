import * as React from 'react';
import ServiceId from '../../../types/ServiceId';
import ServiceInfoBadge from './ServiceInfoBadge';
import ServiceInfoCard from './ServiceInfoCard';
import { Endpoints, Pod, Port, Source, Rule } from '../../../types/ServiceInfo';
import * as API from '../../../services/Api';
import { ToastNotification, ToastNotificationList, Col, Row } from 'patternfly-react';

type ServiceInfoState = {
  labels?: Map<string, string>;
  type: string;
  name: string;
  ip: string;
  ports?: Port[];
  endpoints?: Endpoints[];
  pods?: Pod[];
  rules?: Rule[];
  dependencies?: Source[];
  error: boolean;
  errorMessage: string;
};

class ServiceInfo extends React.Component<ServiceId, ServiceInfoState> {
  constructor(props: ServiceId) {
    super(props);
    this.state = {
      labels: new Map(),
      name: '',
      type: '',
      ip: '',
      ports: [],
      pods: [],
      rules: [],
      dependencies: [],
      error: false,
      errorMessage: ''
    };
  }

  componentWillMount() {
    console.log('Fetching info of a service...');
    API.GetServiceDetail(this.props.namespace, this.props.service)
      .then(response => {
        console.log(response['data']);
        let data = response['data'];
        this.setState({
          labels: data.labels,
          name: data.name,
          type: data.type,
          ports: data.ports,
          endpoints: data.endpoints,
          pods: data.pods,
          dependencies: data.dependencies,
          rules: data.route_rules,
          ip: data.ip
        });
      })
      .catch(error => {
        this.setState({
          error: true,
          errorMessage: 'Could not connect to server'
        });
        console.log(error);
      });
  }

  render() {
    return (
      <div>
        {this.state.error ? (
          <ToastNotificationList>
            <ToastNotification type="danger">
              <span>
                <strong>Error </strong>
                {this.state.errorMessage}
              </span>
            </ToastNotification>
          </ToastNotificationList>
        ) : null}
        <div className="container-fluid container-cards-pf">
          <Row className="row-cards-pf">
            <Col xs={12} sm={12} md={12} lg={12}>
              <ServiceInfoCard
                iconType="pf"
                iconName="service"
                title={this.state.name}
                items={
                  <Row>
                    <Col xs={12} sm={6} md={6} lg={6}>
                      <div className="progress-description">
                        <strong>Labels</strong>
                      </div>
                      {Object.keys(this.state.labels || new Map()).map((key, i) => (
                        <ServiceInfoBadge
                          key={'label_' + i}
                          scale={0.8}
                          style="plastic"
                          color="#0088ce"
                          leftText={key}
                          rightText={this.state.labels ? this.state.labels[key] : ''}
                        />
                      ))}
                      <div className="progress-description">
                        <strong>Type</strong> {this.state ? this.state.type : ''} <strong> Ip</strong>{' '}
                        {this.state ? this.state.ip : ''}
                      </div>
                    </Col>
                    <Col xs={12} sm={6} md={6} lg={6}>
                      <div className="progress-description">
                        <strong>Ports</strong>
                      </div>
                      {(this.state.ports || []).map((port, i) => (
                        <span style={{ marginLeft: '10px' }} key={'port_' + i}>
                          {port.protocol} {port.name} ({port.port})
                        </span>
                      ))}
                    </Col>
                    <Col xs={12} sm={6} md={12} lg={12}>
                      <hr />
                      <div className="progress-description">
                        <strong>Endpoints</strong>
                      </div>
                      {(this.state.endpoints || []).map((endpoint, i) => (
                        <Row key={'endpoint_' + i}>
                          <Col xs={12} sm={6} md={6} lg={6}>
                            <ul>
                              <li style={{ listStyleType: 'none' }}>Addresses</li>
                              <ul>
                                {(endpoint.addresses || []).map((address, u) => (
                                  <li key={'endpoint_' + i + '_address_' + u}>
                                    {address.name} ({address.ip})
                                  </li>
                                ))}
                              </ul>
                            </ul>
                          </Col>
                          <Col xs={12} sm={6} md={6} lg={6}>
                            <ul>
                              <li style={{ listStyleType: 'none' }}>Ports</li>
                              <ul>
                                {(endpoint.ports || []).map((port, u) => (
                                  <li key={'endpoint_' + i + '_port_' + u}>
                                    {port.protocol} {port.name} ({port.port})
                                  </li>
                                ))}
                              </ul>
                            </ul>
                          </Col>
                        </Row>
                      ))}
                    </Col>
                  </Row>
                }
              />
            </Col>
            <Col xs={12} sm={6} md={4} lg={4}>
              <ServiceInfoCard
                iconType="fa"
                iconName="cube"
                title="Pods"
                items={(this.state.pods || []).map((pod, u) => (
                  <div key={'pods_' + u}>
                    <div className="progress-description">{pod['Name']}</div>
                    {Object.keys(pod.labels || new Map()).map((key, i) => (
                      <ServiceInfoBadge
                        key={'pod_labels_badge_' + i}
                        scale={0.8}
                        style="plastic"
                        color="green"
                        leftText={key}
                        rightText={pod.labels ? pod.labels[key] : ''}
                      />
                    ))}
                  </div>
                ))}
              />
            </Col>
            <Col xs={12} sm={6} md={4} lg={4}>
              <ServiceInfoCard
                iconType="pf"
                iconName="route"
                title="Dependencies"
                items={Object.keys(this.state.dependencies || new Map()).map((key, u) => (
                  <div key={'dependencies_' + u}>
                    <div className="progress-description">{key}</div>
                    <ul>
                      {(this.state.dependencies ? this.state.dependencies[key] : []).map((dependency, i) => (
                        <li key={'dependencies_' + u + '_dependency_' + i}>{dependency}</li>
                      ))}
                    </ul>
                  </div>
                ))}
              />
            </Col>
            <Col xs={12} sm={6} md={4} lg={4}>
              <ServiceInfoCard
                iconType="pf"
                iconName="settings"
                title="Services Source"
                items={(this.state.rules || []).map((rule, i) => (
                  <ul style={{ listStyleType: 'none' }} key={'rule' + i}>
                    <li>
                      <strong>Destination</strong> : {rule.destination ? rule.destination['name'] : ''}
                    </li>
                    <li>
                      <strong>Precendence</strong> :{rule.precedence}
                    </li>
                    <li>
                      <strong>Route</strong>:
                      <ul>
                        {(rule.route || []).map((label, u) =>
                          Object.keys(label.labels || new Map()).map((key, n) => (
                            <li key={'rule_' + i + '_label_' + u + '_n_' + n}>
                              {key} : {label.labels[key]}
                            </li>
                          ))
                        )}
                      </ul>
                    </li>
                    {<hr />}
                  </ul>
                ))}
              />
            </Col>
          </Row>
        </div>
      </div>
    );
  }
}

export default ServiceInfo;
