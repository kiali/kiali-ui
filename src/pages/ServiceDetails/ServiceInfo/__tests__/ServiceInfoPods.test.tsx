import * as React from 'react';
import { shallow } from 'enzyme';
import ServiceInfoPods from '../ServiceInfoPods';
import { Pod } from '../../../../types/ServiceInfo';

const pods: Pod[] = [
  {
    name: 'reviews-v2-1234',
    labels: { app: 'reviews', version: 'v2' },
    createdAt: '2018-03-14T10:17:52Z',
    createdBy: []
  },
  {
    name: 'reviews-v3-1234',
    labels: { app: 'reviews', version: 'v3' },
    createdAt: '2018-03-14T10:17:52Z',
    createdBy: []
  },
  {
    name: 'reviews-v1-1234',
    labels: { app: 'reviews', version: 'v1' },
    createdAt: '2018-03-14T10:17:52Z',
    createdBy: []
  }
];

// Mocking "toLocaleString", which is used for dates display
//  since it may produce different results on CI and dev machines, breaking snapshots.
Date.prototype.toLocaleString = jest.fn(Date.prototype.toISOString);

describe('#ServiceInfoPods render correctly with data', () => {
  it('should render service pods', () => {
    const wrapper = shallow(<ServiceInfoPods pods={pods} />);
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
});
