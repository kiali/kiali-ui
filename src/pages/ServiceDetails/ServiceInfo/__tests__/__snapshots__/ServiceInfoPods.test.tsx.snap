// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ServiceInfoPods render correctly with data should render service pods 1`] = `
<div
  className="card-pf"
>
  <Row
    bsClass="row"
    className="row-cards-pf"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      lg={12}
      md={12}
      sm={12}
      xs={12}
    >
      <div
        className="card-pf-body"
        key="pods_0"
      >
        <h3>
          <OverlayTrigger
            defaultOverlayShown={false}
            overlay={
              <Tooltip
                bsClass="tooltip"
                id="pod_names_0"
                placement="right"
                title="Pod Names"
              >
                reviews-v2-1234, reviews-v3-1234, reviews-v1-1234
              </Tooltip>
            }
            placement="top"
            trigger={
              Array [
                "hover",
                "focus",
              ]
            }
          >
            <span>
              reviews-v... (3 replicas)
            </span>
          </OverlayTrigger>
        </h3>
        <div
          className="label-collection"
          key="labels"
        >
          <Label
            key="pod_0_0"
            name="app"
            value="reviews"
          />
        </div>
        <div>
          <span>
            <strong>
              Created at: 
            </strong>
            2018-03-14T10:17:52.000Z
          </span>
        </div>
        <hr />
      </div>
    </Col>
  </Row>
</div>
`;
